distribution/lib/libcdk5/man/cdk_display.3
'\" t
.\" $Id: cdk_display.3,v 1.17 2019/02/20 22:15:59 tom Exp $
.de bP
.ie n  .IP \(bu 4
.el    .IP \(bu 2
..
.ie \n(.g .ds `` \(lq
.el       .ds `` ``
.ie \n(.g .ds '' \(rq
.el       .ds '' ''
.de It
.br
.ie \\n(.$>=3 .ne \\$3
.el .ne 3
.IP "\\$1" \\$2
..
.TH cdk_display 3
.SH NAME
cdk_display - \fBCurses Development Kit\fR Display Capabilities.
.SH SYNOPSIS
Cdk has a number of pre-defined display types.
The following are outlined in this manual page:
.It "\(bu How To Use Colors" 5
.It "\(bu How To Use Different Character Attributes" 5
.It "\(bu How To Justify Strings" 5
.It "\(bu How To Use Special Drawing Characters" 5
.It "\(bu Edit/Display Type Codes (EDisplayType)" 5
.SH DESCRIPTION
Cdk has special formatting commands which can be included in any string which
add highlights, justification, or even colors to a basic string.
These attributes, once set, remain in effect until changed explicitly,
or until the end of the string.
.LP
This manual
page outlines and demonstrates how they work.
.
.SS How To Use Colors
Cdk has the capability to display colors in almost every string type displayed
in a Cdk widget.
.PP
Normally the color pairs are accessed via the COLOR_PAIR macro.
You can still do this, but creating a string with multiple colors is tedious.
That is why the color commands were created.
Use \fBinitCDKColor\fR to create up to 64 color pairs
which you can refer to by number in strings.
.LP
The color settings are stored directly in the string.
When the widget is created or activated, the string is converted
to take advantage of any color commands in the string.
.bP
To turn on a color pair
insert \fI</XX>\fP into the string;
where \fIXX\fR is a numeric value from 0 to the maximum color pair.
.IP
Color pair 0 is the standard default color pair for the screen.
.IP
If you used \fBinitCDKColor\fP, the maximum value for \fIXX\fR is 63
for terminals supporting 8 ANSI colors.
.IP
If you created color pairs directly using \fBinit_pair\fP,
the maximum value for \fIXX\fR is implementation-dependent,
e.g., 63 (for Unix systems) and 255 for ncurses.
.IP
If the terminal does not support color, Cdk uses the \fBbold\fP attribute.
.bP
To turn off a
color pair use the format command <!XX> where \fIXX\fR is a numeric value from
0 to the maximum color pair.
.LP
The following example demonstrates the use of the color commands.
.ne 10
.sp 1
.nf
.ce
\fI----------------------------------------\fR
#include <cdk/cdk.h>

void main()
{
   CDKSCREEN	*cdkscreen;
   CDKLABEL	*demo;
   char		*mesg[4];

   cdkscreen = initCDKScreen (NULL);

   /* Start CDK Colors */
   initCDKColor();

   /* Set the labels up.		*/
   mesg[0] = "</31>This line should have a yellow foreground and a cyan background.<!31>";
   mesg[1] = "</05>This line should have a white  foreground and a blue background.<!05>";
   mesg[2] = "</26>This line should have a yellow foreground and a red  background.<!26>";
   mesg[3] = "<C>This line should be set to whatever the screen default is.";

   /* Declare the labels.	*/
   demo	= newCDKLabel (cdkscreen, CENTER, CENTER, mesg, 4, TRUE, TRUE);

   /* Draw the label		*/
   drawCDKLabel (demo, TRUE);
   waitCDKLabel (demo, ' ');

   /* Clean up			*/
   destroyCDKLabel (demo);
   destroyCDKScreen (cdkscreen);
   endCDK();
   exit (0);
}
.fi
.ce
\fI----------------------------------------\fR
.PP
.
.SS How To Use Different Character Attributes
Cdk also provides attribute commands which allow different character attributes
to be displayed in a Cdk widget.
To use a character attribute the format command
is \fI</X>\fR where \fIX\fR is one of several command characters.
To turn a attribute off use the command \fI<!X>\fR.
The following table outlines the command characters:
.LP
.TS
center tab(/) box;
l l
l l
lw15 lw35 .
\fBCommand Character/Character Attribute\fR
=
B/Bold
U/Underline
K/Blink
R/Reverse
S/Standout
D/Dim
N/Normal
.TE
.LP
The following example demonstrates the use of character display attributes.
.ne 10
.sp 2
.nf
.ce
\fI----------------------------------------\fR
#include <cdk/cdk.h>

void main()
{
   CDKSCREEN    *cdkscreen;
   CDKLABEL     *demo;
   char         *mesg[4];

   cdkscreen = initCDKScreen (NULL);

   /* Start CDK Colors */
   initCDKColor();

   /* Set the labels up.  */
   mesg[0] = "</B/31>Bold text            yellow foreground / blue background.<!31>";
   mesg[1] = "</U/05>Underlined text      white  foreground / blue background.<!05>";
   mesg[2] = "</K/26>Blinking text        yellow foreground / red  background.<!26>";
   mesg[3] = "<C>This line uses the screen default colors.";

   /* Declare the labels.  */
   demo = newCDKLabel (cdkscreen, CENTER, CENTER, mesg, 4, TRUE, TRUE);

   /* Draw the label */
   drawCDKLabel (demo, TRUE);
   waitCDKLabel (demo, ' ');

   /* Clean up */
   destroyCDKLabel (demo);
   destroyCDKScreen (cdkscreen);
   endCDK();
   exit (0);
}
.ce
\fI----------------------------------------\fR
.fi
.LP
Note that color commands and format commands can be mixed inside the same
format marker.
The above example underlines the label marker, which also sets
color pair number 2.
.
.SS How To Justify Strings
Justification commands can
left justify,
right justify,
or center a string of text.
To use a justification format in a string the command <X> is used.
The following table lists the format commands:
.LP
.TS
center tab(/) box;
l l
l l
lw15 lw35 .
\fBCommand/Action.\fR
=
<L>/Left Justified. Default if not stated.
<C>/Centered text.
<R>/Right justified.
<I=X>/Indent the line X characters.
<B=X>/Bullet. X is the bullet string to use.
<F=X>/T{
Links in a file where X is the filename.
This works only with the viewer widget.
T}
.TE

The following example demonstrates how to use the justification commands
in a Cdk widget.
.ce
\fI----------------------------------------\fR
.nf
#include <cdk/cdk.h>

void main()
{
   CDKSCREEN    *cdkscreen;
   CDKLABEL     *demo;
   char         *mesg[5];

   cdkscreen = initCDKScreen (NULL);

   /* Start CDK Colors */
   initCDKColor();

   /* Set the labels up.  */
   mesg[0] = "<R></B/31>This line should have a yellow foreground and a blue background.<!31>";
   mesg[1] = "</U/05>This line should have a white  foreground and a blue background.<!05>";
   mesg[2] = "<B=+>This is a bullet.";
   mesg[3] = "<I=10>This is indented 10 characters.";
   mesg[4] = "<C>This line should be set to whatever the screen default is.";

   /* Declare the labels.  */
   demo = newCDKLabel (cdkscreen, CENTER, CENTER, mesg, 5, TRUE, TRUE);

   /* Draw the label */
   drawCDKLabel (demo, TRUE);
   waitCDKLabel (demo, ' ');

   /* Clean up */
   destroyCDKLabel (demo);
   destroyCDKScreen (cdkscreen);
   endCDK();
   exit (0);
}
.fi
.ce
\fI----------------------------------------\fR
.PP
The bullet format command can take either a single character or a string.
The bullet in the above example would look like
.RS 3
\fI+\fR This is a bullet.
.RE
but if we were to use the following command instead
.RS 3
<B=***>This is a bullet.
.RE
it would look like
.RS 3
\fI***\fR This is a bullet.
.RE
.PP
A format command must be at the beginning of the string.
.
.SS How To Use Special Drawing Characters
Cdk has a set of special drawing characters which can be inserted into any
ASCII file.
These characters are encoded using the format command \*(``<#\fIXX\fP>\*(''
where \fIXX\fP is a two-character name.
The \fBchar2Chtype\fP and \fBchtype2String\fP functions
provide conversion to/from curses \fBchtype\fP data (see \fBcdk_util\fP(3)).
.PP
The following table lists the supported special character commands.
.TS
center tab(/) box;
l l
l l
lw15 lw35 .
\fBSpecial_Character/Character\fR
=
<#UL>/Upper Left Corner
<#UR>/Upper Right Corner
<#LL>/Lower Left Corner
<#LR>/Lower Right Corner
=
<#LT>/Left Tee
<#RT>/Right Tee
<#TT>/Top Tee
<#BT>/Bottom Tee
=
<#HL>/Horizontal Line
<#VL>/Vertical Line
=
<#PL>/Plus Sign
<#PM>/Plus or Minus Sign
<#DG>/Degree Sign
<#CB>/Checker Board
<#DI>/Diamond
<#BU>/Bullet
<#S1>/Scan line 1
<#S9>/Scan line 9
=
<#LA>/Left Arrow
<#RA>/Right Arrow
<#TA>/Top Arrow
<#BA>/Bottom Arrow
.TE
.LP
The character formats can be repeated using an optional numeric repeat value.
To repeat a character add the repeat count within parentheses
to the end of the character format.
The following example draws 10 horizontal-line characters:
.LP
<#HL(10)>
.LP
The following example draws a box within a label window:
.ce
\fI----------------------------------------\fR
.nf
#include <cdk/cdk.h>

void main()
{
   /* Declare variables.  */
   CDKSCREEN    *cdkscreen;
   CDKLABEL     *demo;
   char         *mesg[4];

   cdkscreen = initCDKScreen (NULL);

   /* Start CDK Colors */
   initCDKColor();

   /* Set the labels up.  */
   mesg[0] = "<C><#UL><#HL(26)><#UR>";
   mesg[1] = "<C><#VL></R>This text should be boxed.<!R><#VL>";
   mesg[2] = "<C><#LL><#HL(26)><#LR>";
   mesg[3] = "<C>While this is not.";

   /* Declare the labels.  */
   demo = newCDKLabel (cdkscreen, CENTER, CENTER, mesg, 4, TRUE, TRUE);

   /* Is the label NULL???  */
   if (demo == (CDKLABEL *)NULL)
   {
      /* Clean up the memory.  */
      destroyCDKScreen (cdkscreen);

      /* End curses...  */
      endCDK();

      /* Spit out a message.  */
      printf ("Oops. Can't seem to create the label. Is the window too small?\\n");
      exit (1);
   }

   /* Draw the CDK screen.  */
   refreshCDKScreen (cdkscreen);
   waitCDKLabel (demo, ' ');

   /* Clean up */
   destroyCDKLabel (demo);
   destroyCDKScreen (cdkscreen);
   endCDK();
   exit (0);
}
.fi
.ce
\fI----------------------------------------\fR
.LP
Notice that drawn text can also be justified.
.LP
.
.SS Edit/Display Type Codes (EDisplayType)
.TS
center tab(/) box;
l l
l l
lw15 lw45 .
\fBDisplay_Type/Result\fR
=
vCHAR/Only accepts alphabetic characters.
vLCHAR/T{
Only accepts alphabetic characters.
Maps the character to lower case
when a character has been accepted.
T}
vUCHAR/T{
Only accepts alphabetic characters.
Maps the character to upper case
when a character has been accepted.
T}
vHCHAR/T{
Only accepts alphabetic characters.
Displays a period (\fI.\fR) when a character
has been accepted.
T}
vUHCHAR/T{
Only accepts alphabetic characters.
Displays a period (\fI.\fR) and maps the
character to upper case when a
character has been accepted.
T}
vLHCHAR/T{
Only accepts alphabetic characters.
Displays a period (\fI.\fR) and maps the
character to lower case when a
character has been accepted.
T}
vINT/T{
Only accepts numeric characters.
T}
vHINT/T{
Only accepts numeric characters.
Displays a period (\fI.\fR) when a character
has been accepted.
T}
vMIXED/Accepts any character types.
vLMIXED/T{
Accepts any character types.
Maps the character to lower case
when an alphabetic character has
been accepted.
T}
vUMIXED/T{
Accepts any character types.
Maps the character to upper case
when an alphabetic character has
been accepted.
T}
vHMIXED/T{
Accepts any character types.
Displays a period (\fI.\fR) when a character
has been accepted.
T}
vLHMIXED/T{
Accepts any character types.
Displays a period (\fI.\fR) and maps the
character to lower case when a
character has been accepted.
T}
vUHMIXED/T{
Accepts any character types.
Displays a period (\fI.\fR) and maps the
character to upper case when a
character has been accepted.
T}
vVIEWONLY/Uneditable field.
.TE
.SH SEE ALSO
.BR cdk (3),
.BR cdk_binding (3),
.BR cdk_screen (3),
.BR cdk_util (3)


CHANGES
2024.02.4, released July 13th, 2024

	Important / security related fixes.

	Infrastructure: Ensure a custom tar program specified with
	make TAR=/path/to/custom/tar is used everwhere.

	Defconfig: Pine64 rockpro64: Build ethernet driver in kernel
	rather than as module.

	Updated/fixed packages: apache, composer, daq, fluent-bit,
	ghostscript, glibc, liblockfile, libmad, libopenssl, mesa3d,
	network-manager, ntp, octave, openssh, openvpn, php, python3,
	qt6base, raspberrypi-usbboot, redis, sentry-cli, systemd, tor,
	vlc, xmrig, znc

2024.02.3, released June 10th, 2024

	Important / security related fixes.

	Misc:
	- Fixed a number of compilation issues for hosts using GCC
          14.x, including for kconfig.

	Download:
	- Drop redundant --passive-ftp from default wget options and
	  migrate a number of packages using FTP upstreams to HTTP(S)
	  for compatibility with wget2 which does not have FTP support

	- Ensure downloads are correctly shared for host packages that
	  try to do so using <pkg>_DL_SUBDIR and ensure the correct
	  toplevel directory name is used in generated tarballs when
	  downloads are shared.

	Arch: Enable SSE2 support for x86 Pentium M variant.

	Defconfigs: Orangepi PC/R1, Raspberrypi zero2w, Sipeed
	licheepi nano, Wandboard: Fix compilation issues. PC x86-64
	bios/efi: add wireless regdb to fix kernel warning during
	boot, iMX.8 boards: Adjust FIT external data position to match
	imx-mkimage changes, fixing boot issues.

	Updated/fixed packages: am33x-cm, apcupsd,
	arm-trusted-firmware, binutils-bare-metal, bpftool, catch2,
	ccache, cjson, clamav, cog, coreutils, cryptodev-linux,
	dhcpcd, dmraid, dnsmasq, docker-cli, docker-compose,
	docker-engine, doom-wad, duma, elf2flt, ebtables, elfutils,
	evemu, file, fluidsynth, freeradius-client, freeradius-server,
	freerdp, fwts, gcc, gcc-bare-metal, git, gnuplot, go,
	gr-osmosdr, gst1-plugins-good, hidapi, imagemagick,
	intel-gmmlib, ivi-homescreen, janus-gateway, kexec, kmod,
	kodi, libarchive, libdrm, libedit, libffi, libglib2, libhtp,
	libjwt, libkrb5, libmanette, libmdbx, libmnl,
	libnetfilter_acct, libnetfilter_log, libnpth, libopenssl,
	libpng, libuhttpd, liburiparser, libutempter, libvncserver,
	libwpe, lightning, linux-firmware, linux-pam, lua-cffi,
	luajit, lvm2, lynx, mariadb, mesa3d, micropython, minisign,
	musl, mutt, ncftp, network-manager, newlib-bare-metal, nfacct,
	nfs-utils, octave, openjpeg, openocd, oprofile,
	perl-dbd-mysql, php, postgresql, pppd, privoxy, procps-ng,
	proftpd, proj, putty, python-ml-dtypes, python-pillow,
	python-rds-py, qemu, qt5base, redis-plus-plus, rtl8189fs,
	ruby, samba4, shadow, smcroute, speech, squid, sscep,
	suricata, swupdate, systemd, ti-sgx-km, tinycbor, tinyproxy,
	tor, tpm2-tss, ttyd, uboot-tools, ulogd, vala, weston,
	wpewebkit, x11vnc, zbar, zeromq, zfs, zip, zxing-cpp

	Issues resolved (http://bugs.uclibc.org):

        #15766: LINGMO_PACKAGE_WESTON_XWAYLAND -- 12.0.1 -- fails to build..
	#15973: imx8m bootloader image generation broken since update..
        #16075: Cannot use linux kernel extensions from external tree..
	#16093: kmod and gcc-14 build error

2024.02.2, released May 7th, 2024

	Important / security related fixes.

        Security: The /dev/shm entry in the default /etc/fstab used
        for busybox and sysv init systems was missing the sticky bit,
        allowing users to delete and replace other users files.

        Defconfigs: Mangopi mq1rdw1: Lock kernel headers to 6.6.x.

	Updated/fixed packages: apache, botan,
	brcmfmac_sdio-firmware-rpi, clamav, cmake, containerd,
	crucible, dmenu-wayland, dmraid, docker-cli, docker-engine,
	domoticz, fmc, freeipmi, gdal, gdb, giflib, gnupg2, go,
	google-breakpad, googlefontdirectory, imx-gpu-viv, inih,
	libassuan, libblockdev, libcurl, libdrm, libgpg-error,
	libksba, libnpth, libressl, libtracefs, lxc, mbedtls, mesa3d,
	oatpp, opensc, oprofile, perl-net-ssleay, petitboot,
	php-xdebug, psplash, python-pypng, python-segno,
	python-service-identity, python3, skeleton-init-sysv, systemd,
	tcf-agent, tiff, uuu, wolfssl, xmrig, zic

	Issues resolved (http://bugs.uclibc.org):

        #14416: Mesa3D missing DRI3 support for various Gallium drivers
        #16009: 'select' does not work with 'choice'
        #16036: android-tools version is too old

2024.02.1, released March 26th, 2024

	Important / security related fixes.

	Toolchain: Dropped old Codesourcery ARM / AARCH64 and
	Codescape MIPS external toolchains as they use old Glibc
	versions not compatible with 64-bit time_t

	Defconfigs: Starfive visionfive2: Use correct genimage
	configuration. TI am62x_sk / am64_sk: Use correct optee-os
	platform setting and use a fixed ti-k3-r5-loader version.

	Infrastructure: Use git to fetch CVE data from the FKIE
	repository to work around an issue with missing meta files in
	the releases.

	Updated/fixed packages: axel, bash, bitcoin, botan, busybox,
	containerd, cryptsetup, cvs, docker-compose, docker-engine,
	ell, expat, giflib, gmp, go, harfbuzz, intel-gmmlib,
	intel-mediadriver, iwd, kodi, kodi-pvr-hts, libcgroup, libdrm,
	libgit2, libkrb5, libnfs, libunwind, libvpl,
	libtorrent-rasterbar, libxml2, linux, mariadb, mesa3d,
	mmc-utils, msmtp, multipath-tools, neon, onevpl-intel-gpu,
	openssh, openvpn, osm2pgsql, owfs, php, poco, policycoreutils,
	postgis, privoxy, pulseaudio, pure-ftpd, putty,
	python-configshell-fb, python-rtslib-fb, rauc, restorecond,
	spice, squid, targetcli-fb, tzdata, util-linux, vdr, xmrig,
	xvisor, zic

	Issues resolved (http://bugs.uclibc.org):

	#15997: Build Squid with GCC12.3 in Buildroot 2024.02 will fail

2024.02, released March 5th, 2024

	Various fixes.

	Updated/fixed packages: botan, conmon, dash, efivar, expat,
	fluent-bit, gcc, gst1-vaapi, libcoap, libselinux, liburing,
	libva, libxml2, ltp-testsuite, luvi, newlib-bare-metal,
	nginx-naxsi, octave, openblas, openvmtools, parted, poco,
	powertop, python-aiohttp, python-cheroot, python-django,

mainRelease/lib/libcdk5/cdk_display.c
#include <cdk_int.h>

/*
 * $Author: tom $
 * $Date: 2012/03/20 09:54:37 $
 * $Revision: 1.4 $
 */

/*
 * Given a character pointer, returns the equivalent display type.
 */
EDisplayType char2DisplayType (const char *string)
{
   /* *INDENT-OFF* */
   static const struct {
      const char *name;
      EDisplayType code;
   } table[] = {
      { "CHAR",		vCHAR },
      { "HCHAR",	vHCHAR },
      { "INT",		vINT },
      { "HINT",		vHINT },
      { "UCHAR",	vUCHAR },
      { "LCHAR",	vLCHAR },
      { "UHCHAR",	vUHCHAR },
      { "LHCHAR",	vLHCHAR },
      { "MIXED",	vMIXED },
      { "HMIXED",	vHMIXED },
      { "UMIXED",	vUMIXED },
      { "LMIXED",	vLMIXED },
      { "UHMIXED",	vUHMIXED },
      { "LHMIXED",	vLHMIXED },
      { "VIEWONLY",	vVIEWONLY },
      { 0,		vINVALID },
   };
   /* *INDENT-ON* */

   if (string != 0)
   {
      int n;
      for (n = 0; table[n].name != 0; n++)
      {
	 if (!strcmp (string, table[n].name))
	    return table[n].code;
      }
   }
   return (EDisplayType) vINVALID;
}

/*
 * Tell if a display type is "hidden"
 */
boolean isHiddenDisplayType (EDisplayType type)
{
   boolean result = FALSE;

   switch (type)
   {
   case vHCHAR:
   case vHINT:
   case vHMIXED:
   case vLHCHAR:
   case vLHMIXED:
   case vUHCHAR:
   case vUHMIXED:
      result = TRUE;
      break;
   case vCHAR:
   case vINT:
   case vINVALID:
   case vLCHAR:
   case vLMIXED:
   case vMIXED:
   case vUCHAR:
   case vUMIXED:
   case vVIEWONLY:
      result = FALSE;
      break;
   }
   return result;
}

/*
 * Given a character input, check if it is allowed by the display type,
 * and return the character to apply to the display, or ERR if not.
 */
int filterByDisplayType (EDisplayType type, chtype input)
{
   int result = CharOf (input);

   if (!isChar (input))
   {
      result = ERR;
   }
   else if ((type == vINT ||
	     type == vHINT) &&
	    !isdigit (CharOf (result)))
   {
      result = ERR;
   }
   else if ((type == vCHAR ||
	     type == vUCHAR ||
	     type == vLCHAR ||
	     type == vUHCHAR ||
	     type == vLHCHAR) &&
	    isdigit (CharOf (result)))
   {
      result = ERR;
   }
   else if (type == vVIEWONLY)
   {
      result = ERR;
   }
   else if ((type == vUCHAR ||
	     type == vUHCHAR ||
	     type == vUMIXED ||
	     type == vUHMIXED) &&
	    isalpha (CharOf (result)))
   {
      result = toupper (result);
   }
   else if ((type == vLCHAR ||
	     type == vLHCHAR ||
	     type == vLMIXED ||
	     type == vLHMIXED) &&
	    isalpha (CharOf (result)))
   {
      result = tolower (result);
   }
   return result;
}
